fuse/core.go

```diff
@@ -111,38 +110,51 @@ func (f *fuseNode) applyAttr(a *fuse.Attr, attr *plugin.EntryAttributes) {
 	a.Gid = gid
 }
 
+// Re-discovers the source ancestor of the current node to get fresh data. It returns that ancestor
+// and the path between it and the current node.
+func (f *fuseNode) getSource() (plugin.Parent, []string) {
+	cur, segments := f.parent, []string{plugin.CName(f.entry)}
+	for cur != nil {
+		if plugin.IsPrefetched(cur.entry) {
+			segments = append([]string{plugin.CName(cur.entry)}, segments...)
+			cur = cur.parent
+		} else {
+			// All dirs must contain a parent or they wouldn't have been able to create children.
+			return cur.entry.(plugin.Parent), segments
+		}
+	}
+	return nil, segments
+}
+
+// Re-discovers the current entry, based on any source ancestors.
+func (f *fuseNode) refind(ctx context.Context) (plugin.Entry, error) {
+	parent, segments := f.getSource()
+	if parent == nil {
+		return f.entry, nil
+	}
+	return plugin.FindEntry(ctx, parent, segments)
+}
+
 func (f *fuseNode) Attr(ctx context.Context, a *fuse.Attr) error {
 	// Attr is not a particularly interesting call and happens a lot. Log it to debug like other
 	// activity, but leave it out of activity because it introduces history entries for lots of
 	// miscellaneous shell activity.
 	log.Debugf("FUSE: Attr %v", f)
 
-	var attr plugin.EntryAttributes
-	if f.parent == nil {
-		attr = plugin.Attributes(f.entry)
-	} else {
-		// FUSE caches nodes for a long time, meaning there's a chance
-		// that f's attributes are outdated. CachedList returns the entry's
-		// and its sibling's updated attributes in a single request, so use
-		// it to get f's updated attributes.
-		entries, err := plugin.CachedList(ctx, f.parent)
-		if err != nil {
-			err := fmt.Errorf("could not refresh the attributes: %v", err)
-			activity.Record(ctx, "FUSE: Attr errored %v, %v", f, err)
-			return err
-		}
-		updatedEntry, ok := entries[plugin.CName(f.entry)]
-		if !ok {
-			err := fmt.Errorf("entry does not exist anymore")
-			activity.Record(ctx, "FUSE: Attr errored %v, %v", f, err)
-			return err
-		}
-		attr = plugin.Attributes(updatedEntry)
-		// NOTE: We could set f.entry to updatedEntry, but doing so would require
-		// a separate mutex which may hinder performance. Since updating f.entry
-		// is not strictly necessary for the other FUSE operations, we choose to
-		// leave it alone.
+	// FUSE caches nodes for a long time, meaning there's a chance
+	// that f's attributes are outdated. CachedList returns the entry's
```

Comment:
Should this comment be updated now that we're no longer directly referencing cached list?